File: flutter_native/ios/Classes/DrengrFlutterNativePlugin.swift
import Flutter
import Drengr

/// Bridges the Flutter add-on to the native Drengr iOS SDK. Fail-open throughout.
public class DrengrFlutterNativePlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "drengr_flutter_native",
                                           binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(DrengrFlutterNativePlugin(), channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "start":
            var extra: [String: Any] = [:]
            if let id = args["install_id"] as? String, !id.isEmpty { extra["install_id"] = id }
            if let sid = args["session_id"] as? String, !sid.isEmpty { extra["session_id"] = sid }
            Drengr.start(
                publishableKey: args["publishable_key"] as? String ?? "",
                ingestURL: args["ingest_url"] as? String ?? "",
                appPackage: args["app_package"] as? String ?? "",
                maxBodyBytes: args["max_body_bytes"] as? Int ?? 64 * 1024,
                redactHeaders: Set((args["redact_headers"] as? [String]) ?? []),
                extraContext: extra
            )
            result(true)
        // HttpURLConnection is Android-only.
        case "installUrlConnectionCapture":
            result(false)
        // No set-session API in the native SDK yet; install_id unifies identity.
        case "updateSession":
            result(false)
        case "setEnabled":
            Drengr.setEnabled(args["value"] as? Bool ?? true)
            result(nil)
        case "optOut":
            Drengr.optOut()
            result(nil)
        case "optIn":
            Drengr.optIn()
            result(nil)
        case "identify":
            Drengr.identify(args["external_id"] as? String ?? "",
                            traits: args["traits"] as? [String: Any] ?? [:])
            result(nil)
        case "setExperiment":
            Drengr.setExperiment(args["key"] as? String ?? "",
                                 variant: args["variant"] as? String)
            result(nil)
        case "flush":
            Drengr.flush { DispatchQueue.main.async { result(nil) } }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: flutter_native/android/src/main/kotlin/dev/drengr/flutter_native/DrengrFlutterNativePlugin.kt
package dev.drengr.flutter_native

import android.content.Context
import android.os.Handler
import android.os.Looper
import dev.drengr.sdk.Drengr
import dev.drengr.sdk.DrengrInterceptor
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.Executor

/** Bridges the Flutter add-on to the native Drengr Android SDK. Fail-open throughout. */
class DrengrFlutterNativePlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private var channel: MethodChannel? = null
    private var appContext: Context? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        appContext = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "drengr_flutter_native")
        channel?.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        channel = null
        appContext = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "start" -> result.success(start(call))
                "installUrlConnectionCapture" -> result.success(Drengr.installUrlConnectionCapture())
                // No set-session API in the native SDK yet; install_id unifies identity.
                "updateSession" -> result.success(false)
                "setEnabled" -> {
                    Drengr.setEnabled(call.argument<Boolean>("value") ?: true)
                    result.success(null)
                }
                "optOut" -> { Drengr.optOut(); result.success(null) }
                "optIn" -> { Drengr.optIn(); result.success(null) }
                "identify" -> {
                    Drengr.identify(
                        call.argument<String>("external_id") ?: "",
                        call.argument<Map<String, Any?>>("traits") ?: emptyMap(),
                    )
                    result.success(null)
                }
                "setExperiment" -> {
                    Drengr.setExperiment(
                        call.argument<String>("key") ?: "",
                        call.argument<String>("variant"),
                    )
                    result.success(null)
                }
                "flush" -> Drengr.flush {
                    Handler(Looper.getMainLooper()).post {
                        try { result.success(null) } catch (_: Throwable) {}
                    }
                }
                else -> result.notImplemented()
            }
        } catch (_: Throwable) {
            try { result.success(false) } catch (_: Throwable) {}
        }
    }

    private fun start(call: MethodCall): Boolean {
        val ctx = appContext ?: return false
        val key = call.argument<String>("publishable_key") ?: return false
        val url = call.argument<String>("ingest_url") ?: return false
        val extra = HashMap<String, Any?>()
        call.argument<String>("install_id")?.takeIf { it.isNotEmpty() }?.let { extra["install_id"] = it }
        interceptor = Drengr.start(
            context = ctx,
            publishableKey = key,
            ingestUrl = url,
            appPackage = call.argument<String>("app_package") ?: "",
            maxBodyBytes = (call.argument<Number>("max_body_bytes") ?: DEFAULT_MAX_BODY).toLong(),
            redactHeaders = (call.argument<List<String>>("redact_headers") ?: emptyList()).toSet(),
            extraContext = extra,
        )
        return true
    }

    companion object {
        private const val DEFAULT_MAX_BODY: Long = 64 * 1024

        /** OkHttp has no global hook: host-app native code adds this to its own client builders. */
        @JvmStatic
        @Volatile
        var interceptor: DrengrInterceptor? = null
            private set

        /** Cronet passthrough: a RequestFinishedInfo.Listener, or null when Cronet is absent. */
        @JvmStatic
        fun cronetListener(executor: Executor): Any? = Drengr.cronetListener(executor)
    }
}
